streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala

```diff
@@ -0,0 +1,149 @@
+package org.apache.kafka.streams.scala
+
+import java.lang
+
+import org.apache.kafka.streams.KeyValue
+import org.apache.kafka.streams.kstream.{ForeachAction, GlobalKTable, JoinWindows, Joined, KGroupedStream, KTable, KeyValueMapper, Predicate, Printed, Produced, Serialized, Transformer, TransformerSupplier, ValueJoiner, ValueMapper, ValueMapperWithKey, ValueTransformerSupplier, ValueTransformerWithKeySupplier, KStream => JStream}
+import org.apache.kafka.streams.processor.{ProcessorContext, ProcessorSupplier, TopicNameExtractor}
+import org.apache.kafka.streams.scala.kstream.KStream
+import org.junit.{Assert, Test}
+import org.scalatest.junit.JUnitSuite
+
+class KStreamTest extends JUnitSuite {
+
+  @Test def deprecatedTransformShouldCreateATrueSupplier(): Unit = {
+    val jstream = new MockJStream
+
+    val kstream = new KStream(jstream)
+    //noinspection ScalaDeprecation
+    kstream.transform(new TestTransformer)
+
+    val supplier: TransformerSupplier[String, Long, KeyValue[Int, Double]] = jstream.transformerSupplier.asInstanceOf[TransformerSupplier[String, Long, KeyValue[Int, Double]]]
+    val transformer1 = supplier.get()
+    val transformer2 = supplier.get()
+
+    // we should get two different instances
+    Assert.assertEquals(new KeyValue(1, 1.0), transformer1.transform("A", 1L))
+    Assert.assertEquals(new KeyValue(1, 1.0), transformer2.transform("A", 1L))
+  }
+
+  // demonstrating the condition under which the deprecated method yields incorrect behavior
+  @Test def deprecatedTransformShouldDoesntCreateATrueSupplierWhenPassedAnInstance(): Unit = {
+    val jstream = new MockJStream
+
+    val kstream = new KStream(jstream)
+    val transformer = new TestTransformer
+    //noinspection ScalaDeprecation
+    kstream.transform(transformer)
+
+    val supplier: TransformerSupplier[String, Long, KeyValue[Int, Double]] = jstream.transformerSupplier.asInstanceOf[TransformerSupplier[String, Long, KeyValue[Int, Double]]]
+    val transformer1 = supplier.get()
+    val transformer2 = supplier.get()
+
+    // we should get two different instances
+    Assert.assertEquals(new KeyValue(1, 1.0), transformer1.transform("A", 1L))
+    Assert.assertEquals(new KeyValue(2, 2.0), transformer2.transform("A", 1L))
+  }
+
+
+  private class TestTransformer extends Transformer[String, Long, (Int, Double)] {
+    var transforms: Int = 0
+
+    override def init(context: ProcessorContext): Unit = ???
+
+    override def transform(key: String, value: Long): (Int, Double) = {
+      transforms = transforms + 1
+      (transforms, transforms)
+    }
+
+    override def close(): Unit = ???
+  }
+
+  private class MockJStream extends JStream[String, Long] {
```

Comment:
I tried to use EasyMock, but the capture wouldn't work; I think because of the varargs parameter.